noir-projects/aztec-nr/authwit/src/auth.nr

```diff
@@ -8,16 +8,213 @@ use dep::aztec::protocol_types::{
 };
 use dep::aztec::{prelude::Deserialize, context::{PrivateContext, PublicContext, gas::GasOpts}, hash::hash_args_array};
 
+/**
+ * Authenticaion witness helper library
+ *
+ * Authentication Witness is a scheme for authenticating actions on Aztec, so users can allow third-parties 
+ * (eg protocols or other users) to execute an action on their behalf.
+ * 
+ * This library provides helper functions to manage such witnesses.
+ * The authentication witness, is some "witness" (data) that authenticates a `message_hash`.
+ * The simples example of an authentication witness, is a signature. The signature is the "evidence",
+ * that the signer have seen the message, agrees with it, and have allowed it.
+ * It does not strictly need to be a signature. It could be any kind of "proof" that the message is allowed.
+ * Another proof could be knowing some kind of secret, or having some kind of "token" that allows the message.
+ *
+ * The `message_hash` is a hash of the following structure:
+ * hash(consumer, chain_id, version, inner_hash)
+ * - consumer: the address of the contract that is "consuming" the message,
+ * - chain_id: the chain id of the chain that the message is being consumed on,
+ * - version: the version of the chain that the message is being consumed on,
+ * - inner_hash: the hash of the "inner" message that is being consumed, this is the "actual" message or action.
+ *
+ * While the `inner_hash` could be anything, such as showing your signed a specific message, it will often be 
+ * a hash of the "action" to approve, along with who made the call. As part of this library, we provide a few 
+ * helper functions to deal with "what" these messages are.
+ *
+ * For example, we provide helper functions that is used for checking that the message is encoding the current call.
+ * This can be used to let some contract "allow" another contract to act on its behalf, as long as it can 
+ * show that it is acting on behalf of the contract.
+ *
+ * If we take a case of allowing a contract to transfer tokens on behalf of an account, the `inner_hash` can be 
+ * derived as:
+ * inner_hash = hash(caller, "transfer", hash(to, amount))
+ * 
+ * Where the `caller` would be the address of the contract that is trying to transfer the tokens, and `to` and `amount`
+ * the arguments for the transfer. 
+ * 
+ * Note that we have both a `caller` and a `consumer`, the `consumer` will be the contract that is consuming the message,
+ * in the case of the transfer, it would be the `Token` contract itself, while the caller, will be the actor that is 
+ * allowed to transfer the tokens.
+ *
+ *
+ * The authentication mechanism works differently in public and private contexts. In private, we recall that everything
+ * is executed at the user device, so we can use `oracles` to "ask" the user (not contract) for information. In public
+ * we cannot do this, since it is executed by the sequencer (someone else). Therefore we can instead use a "registry"
```

Comment:
```suggestion
 * we cannot do this, since it is executed by the sequencer (someone else). Therefore we can instead use a public registry contract
```
I would be confused here what is registry and why it means that the data is available to sequencer.